### PR TITLE
Teach Eliza how to remove participants

### PR DESCRIPTION
**Pull request by ryanpbrewster**
@leogrim @camhashemi

### CODE REVIEW

File: kifi-backend/modules/common/app/com/keepit/common/util/DeltaSet.scala [hunk 1]
```diff
@@ -29,6 +29,8 @@ object DeltaSet {
   def addOnly[T](items: Set[T]): DeltaSet[T] = DeltaSet.empty.addAll(items)
   def removeOnly[T](items: Set[T]): DeltaSet[T] = DeltaSet.empty.removeAll(items)
 
+  def simplify[T](ds: DeltaSet[T], xs: Set[T]): DeltaSet[T] = DeltaSet.empty.addAll(ds.added -- xs).removeAll(xs intersect ds.removed)
```
Comment (ryanpbrewster) [inline kifi-backend/modules/common/app/com/keepit/common/util/DeltaSet.scala]: Is it clear what this does?

File: kifi-backend/modules/eliza/app/com/keepit/eliza/model/MessageThread.scala [hunk 2]
```diff
@@ -28,41 +28,22 @@ case class MessageThread(
...
-    this.copy(participants = newParticpiants)
-  }
-
   def withNumMessages(num: Int) = this.copy(numMessages = num)
 
-  def containsUser(user: Id[User]): Boolean = participants.contains(user)
-  def containsNonUser(nonUser: EmailParticipant): Boolean = participants.contains(nonUser)
-  def allParticipants: Set[Id[User]] = participants.allUsers
-  def allEmails: Set[EmailAddress] = participants.allEmails
-
```
Comment (ryanpbrewster) [inline kifi-backend/modules/eliza/app/com/keepit/eliza/model/MessageThread.scala:65]: I strongly prefer having a single way to do things. These methods were unnecessary conveniences.

File: kifi-backend/modules/common/app/com/keepit/common/util/DeltaSet.scala (hunk 1 above)
Comment (leogrim) [inline kifi-backend/modules/common/app/com/keepit/common/util/DeltaSet.scala]: Is this basically removing unnecessary operations from `DeltaSet` when intended to be applied to the `Set`?
I think this is what I got from the formula, but the name didn't really help

File: kifi-backend/modules/eliza/app/com/keepit/eliza/model/MessageThread.scala (hunk 2 above)
Comment (leogrim) [inline kifi-backend/modules/eliza/app/com/keepit/eliza/model/MessageThread.scala:65]: yep much cleaner ✨

File: kifi-backend/modules/eliza/app/com/keepit/eliza/model/MessageThreadParticipants.scala [hunk 3]
```diff
@@ -13,10 +14,17 @@ case class MessageThreadParticipants(userParticipants: Map[Id[User], DateTime],
...
-  def plusUsers(users: Set[Id[User]]) = this.copy(userParticipants = userParticipants -- users)
-  def plusEmails(emails: Set[EmailAddress]) = this.copy(emailParticipants = emailParticipants -- emails.map(EmailParticipant(_)))
-  def minusUsers(users: Set[Id[User]]) = this.copy(userParticipants = userParticipants -- users)
-  def minusEmails(emails: Set[EmailAddress]) = this.copy(emailParticipants = emailParticipants -- emails.map(EmailParticipant(_)))
+  // Note that `diffed` is written such that adding/removing is idempotent. Removing is obvious,
+  // but idempotent addition relies on the right-biased ++ operator on Maps
+  def diffed(users: DeltaSet[Id[User]], emails: DeltaSet[EmailAddress]) = this.copy(
+    userParticipants = users.added.map(_ -> currentDateTime).toMap ++ userParticipants -- users.removed,
+    emailParticipants = emails.added.map(e => EmailParticipant(e) -> currentDateTime).toMap ++ emailParticipants -- emails.removed.map(EmailParticipant(_))
+  )
```
Comment (leogrim) [inline kifi-backend/modules/eliza/app/com/keepit/eliza/model/MessageThreadParticipants.scala:22]: :+1:

File: kifi-backend/modules/common/app/com/keepit/common/util/DeltaSet.scala (hunk 1 above)
Comment (leogrim) [inline kifi-backend/modules/common/app/com/keepit/common/util/DeltaSet.scala]: What's the end goal / benefit of applying this method to a `DeltaSet`?

Comment (ryanpbrewster) [inline kifi-backend/modules/common/app/com/keepit/common/util/DeltaSet.scala]: To simplify situations like `Set(1,2,3) + 3`. In this case, `+3` can be simplified to a no-op.

More concretely, when we modify the participants on a thread we send special notifications to newly added participants. It is suboptimal to send those special notifications to existing participants who have been "erroneously" re-added.

Comment (camhashemi) [inline kifi-backend/modules/common/app/com/keepit/common/util/DeltaSet.scala]: 👍  this is useful across the flow. persisting KeepEvents should be filtered on no-op modifications as well.

**Pull request merged**